File: algorithm_data_structure/leetcode/101.symmetric-tree.py
#
# @lc app=leetcode id=101 lang=python3
#
# [101] Symmetric Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # Use DFS ---> recurision
        def dfs(root1, root2):
            # Check special cases!
            if root1 == root2 == None:
                return True

            if not root1 or not root2:
                return False

            # Start the recursion
            res = (
                root1.val == root2.val
                and dfs(root1.left, root2.right)
                and dfs(root1.right, root2.left)
            )

            return res

        if not root:
            return True
        return dfs(root.left, root.right)


# @lc code=end

File: algorithm_data_structure/leetcode/101.symmetric-tree.cpp
/*
 * @lc app=leetcode id=101 lang=cpp
 *
 * [101] Symmetric Tree
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool helper(TreeNode *left, TreeNode *right)
    {
        // Both nodes are empty! Then it's balance.
        if ((not left) and (not right))
        {
            return true;
        }

        if ((not left) or (not right))
        {
            return false;
        }

        if (left->val != right->val)
        {
            return false;
        }
        // For the symetric: left->left should be compared with right->right.
        // left->right should be compared with right->left.
        return helper(left->left, right->right) and helper(left->right, right->left);
    }

    bool isSymmetric(TreeNode *root)
    {
        if (not root)
        {
            return true;
        }
        else
        {
            return helper(root->left, root->right);
        }
    }
};
// @lc code=end
